### PR TITLE
bugfix: fix MTP k>1 crash by loading embed_tokens weights

### LINKED ISSUE #835: MTP k>1 crashes during warmup due to uninitialized embed_tokens weights
## Description

When using MTP (Multi-Token Prediction) with `num_speculative_tokens > 1` (e.g., k=4), the server crashes during warmup.

## Environment

- Model: GLM-4.7 + GLM-4.7-MTP
- Device: NPU (Ascend)
- Config: `--num_speculative_tokens 4`

## Root Cause

1. **Shared word_embedding causes ATB state corruption**: When k>1, the draft model's `forward()` is called k times per step, each calling the shared `NpuWordEmbeddingImpl`. The layer's internal state gets corrupted during rapid repeated calls.

2. **MTP model doesn't load embed_tokens weights**: `mtp_model_base.h::load_state_dict()` doesn't load `embed_tokens_` weights, even though `mtp_layer_parameters.safetensors` contains `model.embed_tokens.weight`. When we stop sharing the word_embedding layer, the MTP model uses uninitialized weights.

## Crash Log

```
I20260130 01:38:20.409365 profile_manager.cpp:809] Warming up prefill request: tokens=5000
[ERROR] TBE Subprocess[task_distribute] raise error[], main process disappeared!
```

## Stack Trace

```
#0 xllm::layer::NpuWordEmbeddingImpl::build_node_variant_pack()
#1 xllm::layer::NpuWordEmbeddingImpl::forward()
#2 xllm::MtpModelImplBase<xllm::Glm4MoeDecoderLayer>::forward()
```

## Fix

1. Don't share word_embedding layer when `num_speculative_tokens > 1`
2. Add `embed_tokens_` to `load_state_dict()`, `verify_loaded_weights()`, `merge_loaded_weights()` in `mtp_model_base.h`

### PR DESCRIPTION
**Pull request by QwertyJack**
## Summary
- Fix MTP k>1 crash by not sharing word_embedding layer when `num_speculative_tokens > 1`
- Load `embed_tokens_` weights in MTP model's weight loading functions

## Test plan
- [x] Server starts successfully with `--num_speculative_tokens 4`
- [x] Warmup completes without crash
- [x] MTP inference works correctly

Fixes #835

### CODE REVIEW

File: xllm/core/runtime/speculative_worker_impl.cpp [hunk 1]
```diff
@@ -192,8 +192,15 @@ bool SpeculativeWorkerImpl::init_model(const std::string& model_weights_path,
...
-      draft_impl_->set_npu_word_embedding(word_embedding);
+      // NOTE: Do NOT share word_embedding layer when num_speculative_tokens
+      // > 1. Sharing the NpuWordEmbeddingImpl causes ATB operation state
+      // corruption when the layer is called multiple times rapidly in the MTP
+      // k>1 loop. The draft model uses its own word_embedding layer loaded from
+      // its model weights, which is safe for repeated calls.
+      if (options_.num_speculative_tokens() <= 1) {
+        auto word_embedding = impl_->get_npu_word_embedding();
+        draft_impl_->set_npu_word_embedding(word_embedding);
+      }
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply xllm/core/runtime/speculative_worker_impl.cpp:203]: Thank you for the thorough analysis. You raise a valid point about the `lm_head` layer having a similar mutable state pattern.

However, there's a key architectural difference:

1. **`embed_tokens_`**: Constructed in `MtpModelImplBase` constructor but weights weren't loaded → crash when not sharing
2. **`lm_head_`**: Never constructed (`nullptr`) - designed to always use the target model's instance

The MTP model architecture relies on sharing `lm_head` from the target model because:
- MTP only predicts the next hidden states, then uses the shared `lm_head` for logits
- Constructing a separate `lm_head` would require significant refactoring

The crash we observed was specifically in `NpuWordEmbeddingImpl::build_node_variant_pack` during warmup, not in `lm_head`. The `lm_head` sharing has been working in production.

That said, you're correct that this could be a potential issue under high concurrency. I'll note this as a follow-up investigation item if we observe `lm_head`-related crashes in the future.

File: xllm/models/llm/npu/mtp_model_base.h [hunk 2]
```diff
@@ -198,6 +198,8 @@ class MtpModelImplBase : public torch::nn::Module {
       layers_[i]->load_state_dict(
           state_dict.get_dict_with_prefix("layers." + std::to_string(i) + "."));
     }
+    embed_tokens_->load_state_dict(
+        state_dict.get_dict_with_prefix("embed_tokens."));
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply xllm/models/llm/npu/mtp_model_base.h:202]: Thank you for the security analysis.

Regarding `lm_head` weights: The MTP model architecture is designed to **always share** `lm_head` from the target model:

```cpp
// In MtpForCausalLMImplBase:
layer::NpuLmHead lm_head_{nullptr};  // Never constructed

virtual void set_npu_lm_head(layer::NpuLmHead& head) { 
    lm_head_ = head;  // Set from target model
}
```

The `lm_head_` is intentionally left as `nullptr` and populated via `set_npu_lm_head()` from the target model's instance. This is by design because:
1. MTP and target models share the same vocabulary/output projection
2. `lm_head` weights are large (~vocab_size × hidden_size), sharing saves memory

The `shared_head.head.weight` in MTP weights appears to be exported for completeness but is not used in the current architecture.

In contrast, `embed_tokens_` **was constructed** but its weights weren't loaded - that's why we needed to add the weight loading. For `lm_head`, loading weights would first require constructing the layer, which would be a larger architectural change.

File: xllm/models/llm/npu/mtp_model_base.h [hunk 3]
```diff
@@ -198,6 +198,8 @@ class MtpModelImplBase : public torch::nn::Module {
       layers_[i]->load_state_dict(
           state_dict.get_dict_with_prefix("layers." + std::to_string(i) + "."));
     }
+    embed_tokens_->load_state_dict(
```
Comment (yq33victor) [inline xllm/models/llm/npu/mtp_model_base.h:201]: We’ve tested with k > 1 here and found no issues, you might want to check whether something else is causing the problem.
